x509/bash.go

```diff
@@ -12,38 +12,190 @@ import (
+
+if [ $# -ne 3 ] ; then
+	echo "ERROR: $0 <key-create> <cn> <ou>"
+	exit 1
+fi
+key=$1
+cn=$2
+ou=$3
+
+cat >ca.cnf <<EOF
+[req]
+prompt = no
+distinguished_name = dn
+x509_extensions = ext
+
+[dn]
+CN = ${cn}
+OU = ${ou}
+
+[ext]
+keyUsage=critical, keyCertSign
+basicConstraints=critical, CA:TRUE, pathlen:0
+EOF
+
+openssl ecparam -genkey -name prime256v1 | openssl ec -out $key
+openssl req -new -config ca.cnf -key $key
+rm ca.cnf`
+	csrPem := run(script, DeviceCaKeyFile, cn, ou)
+	return signCaCsr("device-ca-*", csrPem)
 }
 
 func SignTlsCsr(csrPem string) string {
-	return run("./"+SignTlsScript, TlsCsrFile)
+	const script = `#!/bin/sh -e
+## This script signs the "tls-csr" returned when creating Factory certificates.
+## This certificate are signed so that the devices trust the TLS connection with the Foundries device gateway.
+
+if [ $# -ne 2 ] ; then
+	echo "ERROR: $0 <tls csr> <tls crt>"
+	exit 1
+fi
+csr=$1
+crt=$2
+
+dns=$(openssl req -text -noout -verify -in $csr | grep DNS:)
+echo "signing with dns name: $dns" 1>&2
+
+cat >server.ext <<EOF
+keyUsage=critical, digitalSignature, keyEncipherment, keyAgreement
+extendedKeyUsage=critical, serverAuth
+subjectAltName=$dns
+EOF
+openssl x509 -req -days 3650 -in $csr -CAcreateserial \
+	-extfile server.ext -CAkey factory_ca.key -CA factory_ca.pem -sha256 -out $crt
+rm server.ext factory_ca.srl || true`
+	return signCsr(script, "tls-*", csrPem)
 }
 
 func SignCaCsr(csrPem string) string {
-	return run("./"+SignCaScript, OnlineCaCsrFile)
+	return signCaCsr("online-ca-*", csrPem)
 }
 
 func SignEl2GoCsr(csrPem string) string {
-	tmpFile, err := os.CreateTemp("", "el2g-*.csr")
+	return signCaCsr("el2g-*", csrPem)
+}
+
+func signCaCsr(tmpFileMask, csrPem string) string {
+	const script = `#!/bin/sh -e
+## This script signs "device ca" signing requests.
+## The request may come from either the Factory (so that lmp-device-register will work)
+## or locally with "create_device_ca" for manufacturing style device creation.
+
+if [ $# -ne 2 ] ; then
+	echo "ERROR: $0 <ca csr> <ca crt>"
+	exit 1
+fi
+csr=$1
+crt=$2
+
+cat >ca.ext <<EOF
+keyUsage=critical, keyCertSign
+basicConstraints=critical, CA:TRUE, pathlen:0
+EOF
+
+if [ -n "$HSM_MODULE" ] ; then
+	lbl=${HSM_TOKEN_LABEL-device-gateway-root}
+	key="pkcs11:token=${lbl};object=root-ca;type=private;pin-value=$HSM_PIN"
+	extra="-CAkeyform engine -engine pkcs11"
+else
+	key=factory_ca.key
+fi
+
+openssl x509 -req -days 3650 $extra -in $csr -CAcreateserial \
+	-extfile ca.ext -CAkey "$key" -CA factory_ca.pem -sha256 -out $crt
+rm ca.ext`
+	return signCsr(script, tmpFileMask, csrPem)
+}
+
+func signCsr(script, tmpFileMask, csrPem string) string {
+	csrFile, err := os.CreateTemp("", tmpFileMask+".csr")
 	subcommands.DieNotNil(err)
-	defer os.Remove(tmpFile.Name())
-	defer tmpFile.Close()
-	_, err = tmpFile.Write([]byte(csrPem))
+	defer os.Remove(csrFile.Name())
+	defer csrFile.Close()
```

Comment:
Maybe a single `defer func() {}` would be slightly better. The commands are printed in the order of execution and the returned error can be verified.

Reply:
I'd better skip this; not sure we are interested in these errors.